### PR TITLE
microshift: avoid oauth and console disruption monitoring

### PR DESCRIPTION
**Pull request by pacevedom**
When https://github.com/openshift/origin/pull/27460 was introduced the tests in MicroShift stopped working because of checking `clusterversion.config.openshift.io`. Since this API group/resource is not available in MicroShift, we need to skip it.

/cc @dgoodwin

### CODE REVIEW

File: test/extended/util/disruption/frontends/known_backends.go
Comment (ingvagabund) [inline test/extended/util/disruption/frontends/known_backends.go:86]: Worth making the function more generic. E.g. `isAPIGroupAvailable(config *rest.Config, apigroup string)`. Or, more preferable to reuse `exutil.DoesApiResourceExist`.

File: test/extended/util/disruption/frontends/known_backends.go
Comment (ingvagabund) [inline test/extended/util/disruption/frontends/known_backends.go:27]: available -> routeAvailable
to align with configAvailable

File: test/extended/util/disruption/frontends/known_backends.go
Comment (ingvagabund) [inline test/extended/util/disruption/frontends/known_backends.go:47]: If `exutil.DoesApiResourceExist` gets used, the oc's config client could be used here instead of creating a new one.

File: test/extended/util/disruption/frontends/known_backends.go
Comment (ingvagabund) [inline test/extended/util/disruption/frontends/known_backends.go:31]: This test originally does not test for `isRouteAvailable` nor for `configAvailable`. To align with the original functionality one should test for both `isRouteAvailable` and `configAvailable` and only then create all the connections and other objects. Or, return error otherwise to fail earlier and avoid initializing half of the functionality.

**Review by ingvagabund — CHANGES_REQUESTED** [review]

**Comment by dgoodwin** [comment]
Apologies for the breakage. Looks good to me will just hold for Jan's requests and then find an approver.

File: test/extended/util/disruption/frontends/known_backends.go
Comment (pacevedom) [reply test/extended/util/disruption/frontends/known_backends.go:86]: Was not keen on refactoring seemingly unrelated parts of the code, but this change might be small enough to do so.

File: test/extended/util/disruption/frontends/known_backends.go
Comment (pacevedom) [reply test/extended/util/disruption/frontends/known_backends.go:27]: Done

File: test/extended/util/disruption/frontends/known_backends.go
Comment (pacevedom) [reply test/extended/util/disruption/frontends/known_backends.go:47]: This would involve creating an oc object here only to use the config client, is that an overkill? oc objects do many operations besides creating clients for APIs.

File: test/extended/util/disruption/frontends/known_backends.go
Comment (pacevedom) [reply test/extended/util/disruption/frontends/known_backends.go:31]: The routes used for monitoring belong to the console and the oauth api server, which are part of a regular OCP cluster. Since we don't have them in MicroShift this should be good enough to determine whether disruption tests should be created.
The one over the configAvailable comes from [this PR](https://github.com/openshift/origin/pull/27460), which introduces capability checking and a fallback if capability checking fails (which means we should still create the disruption monitor). Since we don't have the `clusterversion` in MicroShift we need to skip it.

**Comment by pacevedom** [comment]
> Apologies for the breakage. Looks good to me will just hold for Jan's requests and then find an approver.

No worries! All this stuff is pretty new so it was difficult to know if there is something broken :)

File: test/extended/util/disruption/frontends/known_backends.go
Comment (ingvagabund) [reply test/extended/util/disruption/frontends/known_backends.go:47]: If that's the case I am taking my comment back

File: test/extended/util/disruption/frontends/known_backends.go
Comment (ingvagabund) [reply test/extended/util/disruption/frontends/known_backends.go:86]: I meant to refactor `isConfigAPIGroupAvailable`. Or call `exutil.DoesApiResourceExist` as it is instead of `isConfigAPIGroupAvailable`. Sorry for the confusion.

File: test/extended/util/disruption/frontends/known_backends.go
Comment (ingvagabund) [reply test/extended/util/disruption/frontends/known_backends.go]: What I meant was:
```go
	// Ingress monitoring checks for oauth and console routes to monitor healthz endpoints. Check availability
	// before setting up any monitors.
	routeAvailable, err := isRouteAvailable(ctx, clusterConfig, oauthRouteNamespace, oauthRouteName)
	if err != nil {
		return err
	}
	configAvailable, err := isConfigAPIGroupAvailable(clusterConfig)
	if err != nil {
		return err
	}

	if !routeAvailable || !configAvailable {
		return fmt.Error("Route or config not available")
	}
	if routeAvailable {
		if err := createOAuthRouteAvailableWithNewConnections().StartEndpointMonitoring(ctx, m, nil); err != nil {
			return err
		}
		if err := createOAuthRouteAvailableWithConnectionReuse().StartEndpointMonitoring(ctx, m, nil); err != nil {
			return err
		}
	}
	if configAvailable {
		// Some jobs explicitly disable the console and other features. Check if it's disabled and if so,
		// do not run a disruption monitoring backend for it.
		configClient, err := configclient.NewForConfig(clusterConfig)
...
```

File: test/extended/util/disruption/frontends/known_backends.go
Comment (pacevedom) [reply test/extended/util/disruption/frontends/known_backends.go]: Done

File: test/extended/util/disruption/frontends/known_backends.go
Comment (pacevedom) [reply test/extended/util/disruption/frontends/known_backends.go]: The problem with that check is returning the error. This would prevent tests from running in MicroShift, as it fails to prepare before launching any of them.
If there are no routes for oauth and/or console then there is no point in setting anything up, as those services might not be installed/active, and this should mean no error.
It is a monitor device watching for any possible disruptions during the whole test run. If there are no services to watch, there cant be any disruptions, therefore the prepare phase should progress and setup nothing.

Comment (ingvagabund) [reply test/extended/util/disruption/frontends/known_backends.go]: > If there are no routes for oauth and/or console then there is no point in setting anything up, as those services might not be installed/active, and this should mean no error.

Valid point.

> The problem with that check is returning the error. This would prevent tests from running in MicroShift, as it fails to prepare before launching any of them.

I was under assumption running `StartEndpointMonitoring` requires existence of the route. Though, checking the code it seems it does not. `disruptionSampler.produceSamples` will produce an error over non-existing route connection check. Though, the monitor will keep running and collecting samples. Seems fair to skip running `StartEndpointMonitoring` for both cases when there's no OAuth route.

File: test/extended/util/disruption/frontends/known_backends.go
Comment (ingvagabund) [inline test/extended/util/disruption/frontends/known_backends.go:31]: Is there any OpenShift installation which does not deploy routes? Given routes are deployed by default and MicroShift deploys them as well, why to perform the check?

Comment (pacevedom) [reply test/extended/util/disruption/frontends/known_backends.go:31]: There are two routes in OCP, which come from the oauth api server and the console operator. The former is always installed as it is a core operator. The latter can be disabled in 4.12+.
Neither of those exist in MicroShift, so the routes wont exist unless they are introduced in the future. In an OCP deployment there should always be (at least) a route for the oauth api server.

**Comment by ingvagabund** [comment]
/hold
Let's wait for https://github.com/openshift/origin/pull/27416 to merge

**Pull request merged**
